### ISSUE TITLE
Removing Multiplayer Component Causes Build Error

### ISSUE THREAD

**Issue by AMZN-Gene** [o3de/o3de#12973]
If you remove an auto-component XML, AZCodeGen won't remove the generated code file and it causes a compile error.
Users has to manually find and delete the generated file inside the build folder.

**Steps to reproduce**
1. Create a new multiplayer autocomponent xml and add it to cmake files. (Set ComponentOverride and ControllerOverride to false, so you dont need any custom code file)
2. Compile (see that compile works)
3. Delete the the autocomponent and remove it from cmake file list.
4. Compile

**Expected behavior**
Code compiles successfully.

**Actual behavior**
Build error occurs. User has to manually delete the generated code file and rebuild

```
Build started...
5>------ Build started: Project: Multiplayer.Static, Configuration: profile x64 ------
5>unity_2_cxx.cxx
5>Multiplayer.Static.vcxproj -> D:\prj\MultiplayerSample\build\windows\lib\profile\Multiplayer.Static.lib
6>------ Build started: Project: MultiplayerSample.Static, Configuration: profile x64 ------
7>------ Build started: Project: Multiplayer.Debug, Configuration: profile x64 ------
8>------ Build started: Project: Multiplayer (Gems\Multiplayer\Code\Multiplayer), Configuration: profile x64 ------
6>Running AutoGen for MultiplayerSample.Static
6>*.AutoComponent.xml,AutoComponent_Header.jinja,$path/$fileprefix.AutoComponent.h : error Processing expansion rule
6>An unexpected error occurred, please report the error you encountered and include your build output
6>Traceback (most recent call last):
6>  File "D:\o3de\cmake\AzAutoGen.py", line 130, in ProcessTemplateConversion
6>    xml = etree.parse(dataInputFile)
6>  File "D:\o3de\python\runtime\python-3.10.5-rev1-windows\python\lib\xml\etree\ElementTree.py", line 1224, in parse
6>    tree.parse(source, parser)
6>  File "D:\o3de\python\runtime\python-3.10.5-rev1-windows\python\lib\xml\etree\ElementTree.py", line 569, in parse
6>    source = open(source, "rb")
6>FileNotFoundError: [Errno 2] No such file or directory: 'D:\\prj\\MultiplayerSample\\Gem\\Code\\Source\\AutoGen\\SomeNewComponent.AutoComponent.xml'
6>
6>During handling of the above exception, another exception occurred:
6>
6>Traceback (most recent call last):
6>  File "D:\o3de\cmake\AzAutoGen.py", line 419, in <module>
6>    autoGenResult = ExecuteExpansionRules(autogenConfig, dataInputSet, outputFiles)
6>  File "D:\o3de\cmake\AzAutoGen.py", line 374, in ExecuteExpansionRules
6>    ProcessExpansionRule(autogenConfig, sourceFiles, templateFiles, templateCache, expansionRule, dataInputSet, outputFiles)
6>  File "D:\o3de\cmake\AzAutoGen.py", line 332, in ProcessExpansionRule
6>    ProcessTemplateConversion(autogenConfig, dataInputSet, dataInputFiles, templateFile, outputFileAbsolute, templateCache)
6>  File "D:\o3de\cmake\AzAutoGen.py", line 167, in ProcessTemplateConversion
6>    PrintError('%s(%s) : %s' % (fileinput.filename(), str(fileinput.filelineno()), e.message))
6>  File "D:\o3de\python\runtime\python-3.10.5-rev1-windows\python\lib\fileinput.py", line 121, in filename
6>    raise RuntimeError("no active input()")
6>RuntimeError: no active input()
6>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(241,5): error MSB8066: Custom build for 'D:\prj\MultiplayerSample\build\windows\CMakeFiles\c89cbd3be96997261947eb8d2de4f388\SomeNewComponent.AutoComponent.h.rule;D:\prj\MultiplayerSample\Gem\Code\CMakeLists.txt' exited with code 1.
6>Done building project "MultiplayerSample.Static.vcxproj" -- FAILED.
9>------ Build started: Project: MultiplayerSample (Gem\Code\MultiplayerSample), Configuration: profile x64 ------
9>Building Custom Rule D:/prj/MultiplayerSample/Gem/Code/CMakeLists.txt
9>CMake is re-running because D:/prj/MultiplayerSample/build/windows/External/MultiplayerSample-512ed798/Gem/Code/CMakeFiles/generate.stamp is out-of-date.
9>  the file 'D:/prj/MultiplayerSample/Gem/Code/multiplayersample_files.cmake'
9>  is newer than 'D:/prj/MultiplayerSample/build/windows/External/MultiplayerSample-512ed798/Gem/Code/CMakeFiles/generate.stamp.depend'
9>  result='-1'
9>-- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19044.
9>-- Using package D:/.o3de/3rdParty/packages/zlib-1.2.11-rev5-windows
9>-- Configuring done
9>-- Generating done
9>-- Build files have been written to: D:/prj/MultiplayerSample/build/windows
9>   Creating library D:/prj/MultiplayerSample/build/windows/lib/profile/MultiplayerSample.lib and object D:/prj/MultiplayerSample/build/windows/lib/profile/MultiplayerSample.exp
9>MultiplayerSample.vcxproj -> D:\prj\MultiplayerSample\build\windows\bin\profile\MultiplayerSample.dll
10>------ Build started: Project: MultiplayerSample.ServerLauncher, Configuration: profile x64 ------
10>Copying MultiplayerSample.ServerLauncher runtime dependencies to output...
10>MultiplayerSample.ServerLauncher.vcxproj -> D:\prj\MultiplayerSample\build\windows\bin\profile\MultiplayerSample.ServerLauncher.exe
========== Build: 9 succeeded, 1 failed, 112 up-to-date, 0 skipped ==========
```

**Found in Branch**
development latest October 28 2022

**Desktop/Device (please complete the following information):**
 Windows VS 2019

**Comment by lmbr-pip** [comment]
Removed sig/network from issue. Unsure AzAutoGen ownership. Asking SIG/Core or SIG/Build for input as potential owners of AzAutoGen.

**Comment by lumberyard-employee-dm** [comment]
This behavior is expected as the AutoGen system globs the list of files using a wildcard.
So when the an AutoComponent.xml file is removed, it doesn't know that it needs to CMake reconfigure.

The list of `DEPENDS` in the custom command in [LyAutoGen.cmake](https://github.com/o3de/o3de/blob/development/cmake/LyAutoGen.cmake#L44) would need to have the specific file on the list

**Comment by AMZN-Gene** [comment]
Dang, even though I remove SomeComponent.AutoComponent.xml from the projects cmake list? Go ahead and close if it's expected behavior, but I wouldn't guess a new user knowing this is expected.

**Comment by AMZN-Gene** [comment]
> This behavior is expected as the AutoGen system globs the list of files using a wildcard. So when the an AutoComponent.xml file is removed, it doesn't know that it needs to CMake reconfigure.
> 
> The list of `DEPENDS` in the custom command in [LyAutoGen.cmake](https://github.com/o3de/o3de/blob/development/cmake/LyAutoGen.cmake#L44) would need to have the specific file on the list

Bumpping since I keep running into this >.<. 
I thought editing any __files.cmake_ would cause a re-configure and thus a re-generation of autocomponents. 
Just to be sure, @lumberyard-employee-dm suggestion is that I could edit [LyAutoGen.cmake](https://github.com/o3de/o3de/blob/development/cmake/LyAutoGen.cmake#L44) to re-generate if some_particular_files.cmake is altered?

**Comment by AMZN-Gene** [comment]
_"Editing any file that cmake includes would cause a re-configure, but that would be dependent on the file being included by cmake."_
Sorry, my confusion is that editing cmake causes a re-configure but not a re-generate? Whenever I delete _mycomponent.autogen.XML_ I also remove the _mycomponent.autogen.XML_ from the cmake list. But this doesn't cause auto-gen code to be re-gen'd. Or do you mean in order auto-gen to delete the generated files like _mycomponent.autogen.cpp_ and _mycomponent.autogen.h_, then _mycomponent.autogen.cpp_ and _mycomponent.autogen.h_ needs to appear in a cmake file?

**Comment by lumberyard-employee-dm** [comment]
CMake is re-running the configure step followed by it's generation steps.
But the generation of the codegen files is happening through AzAutoGen.py script that is invoked by the custom command .

CMake doesn't delete the generated files because it is not CMake responsibility, it can only be told that they file is an [output](https://github.com/o3de/o3de/blob/development/cmake/LyAutoGen.cmake#L43) of a custom command. Based on CMake knowing about the file it can re-run CMake if the generated file is removed by the user.

However once CMake re-runs, it will run AzAutoGen.py and if the `mycomponent.autogen.xml` no longer exists, then a new list of generated [outputs](https://github.com/o3de/o3de/blob/development/cmake/LyAutoGen.cmake#L43) will be given to CMake.
The old generated `mycomponent.autogen.cpp` and `mycomponent.autogen.h` will no longer be tracked for changes.

All that being said, unless `AzAutoGen.py` is deleting files that it doesn't generated from the generation directory, the files will not go away.

**Comment by lumberyard-employee-dm** [comment]
> > _"Editing any file that cmake includes would cause a re-configure, but that would be dependent on the file being included by cmake."_
> > Sorry, my confusion is that editing cmake causes a re-configure but not a re-generate? Whenever I delete _mycomponent.autogen.XML_ I also remove the _mycomponent.autogen.XML_ from the cmake list. But this doesn't cause auto-gen code to be re-gen'd. Do you mean in order auto-gen to delete the generated files like _mycomponent.autogen.cpp_ and _mycomponent.autogen.h_, then _mycomponent.autogen.cpp_ and _mycomponent.autogen.h_ needs to appear in a cmake file?
> 
> CMake is re-running the configure step followed by it's generation steps. But the generation of the codegen files is happening through AzAutoGen.py script that is invoked by the custom command .
> 
> CMake doesn't delete the generated files because it is not CMake responsibility, it can only be told that they file is an [output](https://github.com/o3de/o3de/blob/development/cmake/LyAutoGen.cmake#L43) of a custom command. Based on CMake knowing about the file it can re-run CMake if the generated file is removed by the user.
> 
> However once CMake re-runs, it will run AzAutoGen.py and if the `mycomponent.autogen.xml` no longer exists, then a new list of generated [outputs](https://github.com/o3de/o3de/blob/development/cmake/LyAutoGen.cmake#L43) will be given to CMake. The old generated `mycomponent.autogen.cpp` and `mycomponent.autogen.h` will no longer be tracked for changes.
> 
> All that being said, unless `AzAutoGen.py` is deleting files that it doesn't generated from the generation directory, the files will not go away.

Pretty much if for example the AzAutoGen outputs generated files to the folder of `D:/o3de/build/windows/External/Gems/Multiplayer-cedbd128/Code/Azcg/Generated/Source/Autogen` folder. 
To keep the example simple lets focus on a 1 xml file to 1 output file generation: i.e if AzAutoGen.py was provided an `<Autogen>.xml`, it would produce `<Autogen>.cpp`

If there are two input files of `AutoComponent.xml` and `Multiplayer.AutoPackets.xml`, then running those files through AzAutoGen.py would produce the following two files
|Original Generated Files|
|---|
|`D:/o3de/build/windows/External/Gems/Multiplayer-cedbd128/Code/Azcg/Generated/Source/Autogen/AutoComponent.cpp`|
|`D:/o3de/build/windows/External/Gems/Multiplayer-cedbd128/Code/Azcg/Generated/Source/Autogen/Multiplayer.AutoPackets.cpp`|

Afterwards the 'AutoComponent.xml` is removed by the author from the list of input files to use for generation.
When AzAutoGen.py reruns, it produces only a single generated file.

|Second Generated Files|
|---|
|`D:/o3de/build/windows/External/Gems/Multiplayer-cedbd128/Code/Azcg/Generated/Source/Autogen/Multiplayer.AutoPackets.cpp`|

At this point the old `D:/o3de/build/windows/External/Gems/Multiplayer-cedbd128/Code/Azcg/Generated/Source/Autogen/AutoComponent.cpp` that was based on the now removed `AutoComponent.xml` still exist within the generated directory.

What could be done by AzAutoGen.py is that can check the generated directory for any files that it did not generate during its current run and remove them after generation.

AzAutoGen.py populates the list output files in its [ProcessExpansionRule](https://github.com/o3de/o3de/blob/development/cmake/AzAutoGen.py#L321-L344) method. 
With that list it can cross check that against all files in the generation directory and files not in the generation list can then be removed.

**Comment by AMZN-Gene** [comment]
Ah that makes sense, AzCodeGen is simply generating outputs; it keeps no history record of what caused the output to begin with and so doesn't ever delete any generated files. I like the idea of cross-checking the list of output files and deleting anything that isn't in the generation list. That would definitely work for multiplayer's use of auto-gen.

**Comment by lumberyard-employee-dm** [comment]
> Ah that makes sense about cmake not being responsible for generating, it just has a list of files it should know about. On the other hand AzCodeGen is simply generating outputs; it keeps no history record of what caused the output to begin with and so doesn't ever delete any generated files. I like the idea of cross-checking the list of output files and deleting anything that isn't in the generation list. That would definitely work for multiplayer's use of auto-gen; I might have a poke at it. Thanks!

If you do that can put that option behind a `--prune` option and we can update the invocation in the `LyAutoGen.cmake` to add that parameter

**Comment by kberg-amzn** [comment]
@lumberyard-employee-dm I still don't fully understand why this causes a build issue.
1. We remove the autocomponent.xml from the cmake lists
2. That causes a cmake reconfigure which will query autogen for a new set of generated files and dependencies
3. The projects emitted by cmake should now no longer include anything related to the autocomponent.xml
Sure, the stale generated artefact files may still be kicking around on disk, but they should now be completely unused. Is cmake skipping the configure step to deduce the set of generated files?

**Comment by lumberyard-employee-dm** [comment]
@gene-walt I don't believe this is causing a build issue correct? 
Isn't your issue related to the stale generated artifact files being left on disk.

**Comment by AMZN-Gene** [comment]
> @gene-walt I don't believe this is causing a build issue correct?
> Isn't your issue related to the stale generated artifact files being left on disk?

It causes a build error. Compilation fails.

**Comment by lumberyard-employee-dm** [comment]
> > @gene-walt I don't believe this is causing a build issue correct?
> > Isn't your issue related to the stale generated artifact files being left on disk?
> 
> It causes a build error. Compilation fails.

So let's go back a bit.
I have re-assessed the situation.

An `*.autocomponent.xml `is removed from a CMakeLists.txt. 
That *.autocomponent.xml gets added as a dependency of the [AzAutoGen](https://github.com/o3de/o3de/blob/development/cmake/LyAutoGen.cmake#L42-L48) custom_command. 
See the `DEPENDS ${AZCG_DEPENDENCIES}` line.

That indicates to CMake that the build project generated by CMake should re-run that command when one of those files are modified or remove.
But that does **NOT** indicate that that CMake configure should re-run if that file is modified or removed.
For example a modifying or removing `Entity.cpp` will not cause CMake configure to re-run, because that is only a file dependency. But modifying or removing the `AzCore CMakeLists.txt` or any other file that was explicitly added to CMake via an `include()` or an `add_subdirectory()` call will re-run CMake configure as that adds a configure dependency.

Now the `*.AutoComponent..xml`, `*.AutoComponent.jinja` files are only file dependencies as state per the [add_custom_command](https://cmake.org/cmake/help/latest/command/add_custom_command.html?highlight=add_custom_command#generating-files) DEPENDS documentation.
> DEPENDS
>
>    Specify files on which the command depends. Each argument is converted to a dependency as follows:
>
> 1. If the argument is the name of a target (created by the [add_custom_target()](https://cmake.org/cmake/help/latest/command/add_custom_target.html#command:add_custom_target), [add_executable()](https://cmake.org/cmake/help/latest/command/add_executable.html#command:add_executable), or [add_library()](https://cmake.org/cmake/help/latest/command/add_library.html#command:add_library) command) a target-level dependency is created to make sure the target is built before any target using this custom command. Additionally, if the target is an executable or library, a file-level dependency is created to cause the custom command to re-run whenever the target is recompiled.
>
> 2. If the argument is an **absolute path, a file-level dependency** is created on that path.
>
> 3. If the argument is the name of a source file that has been added to a target or on which a source file property has been set, a file-level dependency is created on that source file.
>
>  4. If the argument is a relative path and it exists in the current source directory, a file-level dependency is created on that file in the current source directory.
>
>  5. Otherwise, a file-level dependency is created on that path relative to the current binary directory.

Now what we want is a **configure dependency**.

That can be added via the CMake Directory property of [CMAKE_CONFIGURE_DEPENDS](https://cmake.org/cmake/help/latest/prop_dir/CMAKE_CONFIGURE_DEPENDS.html?highlight=configure_depends)

We have an example of adding a CMake configure dependency in FileUtil.cmake [ly_read_file()](https://github.com/o3de/o3de/blob/development/cmake/FileUtil.cmake#L133-L141) which you can examine.

We just need to add to APPEND to the CMAKE_CONFIGURE_DEPENDS property the list of `*.AutoComponent.xml` and `*.AutoComponent*.jinja` files in the `ly_add_autogen` cmake function after we gather complete the AZCG_DEPENDENCIES list of [here](https://github.com/o3de/o3de/blob/development/cmake/LyAutoGen.cmake#L42-L48).


TLDR: To fix this issue, set the AZCG_DEPENDENCIES as a CMake configure dependency right after this line : https://github.com/o3de/o3de/blob/development/cmake/LyAutoGen.cmake#L41

```
set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS ${AZCG_DEPENDENCIES})
```
